### PR TITLE
Extensibility and library updates

### PR DESCRIPTION
**Pull request by brh**
Thanks for this library it has been of great use to my team.  But our foray into a [componentized architecture](https://github.com/julianomoraes/componentizationArch) we feel we need the ability to pass a factory to the map function.

### CODE REVIEW

File: fast-list/src/main/java/com/list/rados/fast_list/BaseList.kt [hunk 1]
```diff
@@ -52,15 +52,24 @@ fun <T> RecyclerView.update(newItems: List<T>) {
 open class FastListAdapter<T>(private var items: MutableList<T>, private var list: RecyclerView
 ) : RecyclerView.Adapter<FastListViewHolder<T>>() {
 
-    private inner class BindMap(val layout: Int, var type: Int = 0, val bind: View.(item: T) -> Unit, val predicate: (item: T) -> Boolean)
+    private inner class BindMap(val layout: Int?, var type: Int = 0, val bind: View.(item: T) -> Unit, val predicate: (item: T) -> Boolean) {
```
Comment (radoyankov) [inline fast-list/src/main/java/com/list/rados/fast_list/BaseList.kt]: Why did you change the layout to be nullable? If you keep it as an Int, you won't have to add the !! down the line.

File: fast-list/src/main/java/com/list/rados/fast_list/BaseList.kt [hunk 2]
```diff
@@ -52,15 +52,24 @@ fun <T> RecyclerView.update(newItems: List<T>) {
...
 
     private var bindMap = mutableListOf<BindMap>()
     private var typeCounter = 0
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FastListViewHolder<T> {
         return bindMap.first { it.type == viewType }.let {
-            FastListViewHolder(LayoutInflater.from(parent.context).inflate(it.layout,
-                    parent, false), viewType)
+            if (it.layoutFactory == null) {
+                return FastListViewHolder(LayoutInflater.from(parent.context).inflate(it.layout!!,
```
Comment (radoyankov) [inline fast-list/src/main/java/com/list/rados/fast_list/BaseList.kt]: Try to not have the !! conversion here.

**Comment by radoyankov** [comment]
Those are some interesting additions. I've added some comments, and if you take a look at them, I'll merge it and update the library with these changes.

**Comment by brh** [comment]
Ok, I have updated the pull requests with the suggested changes.

**Comment by radoyankov** [comment]
There is still a place where you convert a nullable to a non nullable ("!!"), at line 71 in BaseList. You can remove that as well. Other than that it looks good.

**Comment by brh** [comment]
Had to change if to get rid of !!, still fairly new to Kotlin
